api/fma/v1alpha1/launcherpoolpolicy_types.go

```diff
@@ -0,0 +1,220 @@
+//
+// 2. **Zero Matching CountForLauncher**
+//   - When no CountForLauncher matches a specific (Node, Accelerator, LauncherConfig) tuple:
+//     - No pre-provisioning occurs for that specific combination
+//     - Launcher pods are created on-demand when needed
+//
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:shortName=lpp
+
+type LauncherPoolPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   LauncherPoolPolicySpec   `json:"spec,omitempty"`
+	Status LauncherPoolPolicyStatus `json:"status,omitempty"`
+}
+
+// LauncherPoolPolicySpec defines the node-level idle pool configuration.
+type LauncherPoolPolicySpec struct {
+	// LauncherPoolForNodeType defines pool spec per node type.
+	LauncherPoolForNodeType []LauncherPoolForNodeType `json:"launcherPoolForNodeType"`
+}
+
+// LauncherPoolForNodeType defines launcher count for a class of nodes.
+type LauncherPoolForNodeType struct {
+	// Selector describes the hardware characteristics of target nodes.
+	// +kubebuilder:validation:Required
+	// Introduce an EnhancedNodeSelector that supports combining label-based
+	// matching with resource field conditions.
+	// For example:
+	// enhancedNodeSelector:
+	//  # 1. Label selector (compatible with existing metav1.LabelSelector)
+	//  labelSelector:
+	//    matchLabels:
+	//      nvidia.com/gpu.family: ada-lovelace
+	//    matchExpressions:
+	//      - key: node.kubernetes.io/instance-type
+	//        operator: In
+	//        values: ["gx3-48x240x2l40s", "gx3-96x480x4l40s"]
+	//
+	//  # 2. Resource condition selector (new capability)
+	//  resourceRequirements:
+	//    allocatable:
+	//      cpu:
+	//        min: "16"
+	//        max: "64"
+	//      memory:
+	//        min: "128Gi"
+	//        max: "512Gi"
+	//      gpus:
+	//        nvidia.com/gpu:
+	//          min: "2"
+	//	        max: "8"
+	//  acceleratorSelector:
+	//    type: "nvidia.com/gpu"
+	//    memory: "48Gi"
+	//    count: 4
+	EnhancedNodeSelector EnhancedNodeSelector `json:"enhancedNodeSelector"`
+
+	// PerAcceleratorCount defines pre-configuration quantities for each accelerator type
+	PerAcceleratorCount []PerAcceleratorCount `json:"perAcceleratorCount"`
+}
+
+// PerAcceleratorCount defines configuration for specific accelerators
+type PerAcceleratorCount struct {
+	// AcceleratorSelector accelerator selector
+	AcceleratorSelector AcceleratorSelector `json:"acceleratorSelector"`
+
+	// CountForLauncher is the total number of launcher for each LauncherConfig
+	// to maintain on each matching node per accelerator.
+	// If two different counts are specified for the same (Node, Accelerator, LauncherConfig),
+	// the higher count is used and will be populated into LauncherPoolPolicyStatus.
+	// If no CountForLauncher applies to a given (Node, Accelerator, LauncherConfig), this Node
+	// will be ignored for this LauncherConfig.
+	// +kubebuilder:validation:MinItems=1
```

Comment:
Why must this slice be non-empty, while the other slices have no such requirement?

The common pattern in Kubernetes APIs is to allow an empty slice and include `,omitempty` in the json tag on the field. I think that would be OK here too, although it is also true that an empty slice means the whole struct contributes nothing. On the third hand, forbidding an empty slice is a bit of irregularity: for any length larger than one, removing an entry is allowed, but for length == 1, removing an entry is not allowed.

Reply:
that's fine for me to remove this constraints